src/Features/Core/Portable/MoveToNamespace/AbstractMoveToNamespaceCodeAction.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MoveToNamespace
+{
+    internal abstract partial class AbstractMoveToNamespaceCodeAction : CodeActionWithOptions
+    {
+        private readonly IMoveToNamespaceService _changeNamespaceService;
+        private readonly MoveToNamespaceAnalysisResult _moveToNamespaceAnalysisResult;
+
+        public AbstractMoveToNamespaceCodeAction(IMoveToNamespaceService changeNamespaceService, MoveToNamespaceAnalysisResult analysisResult)
+        {
+            _changeNamespaceService = changeNamespaceService;
+            _moveToNamespaceAnalysisResult = analysisResult;
+        }
+
+        public override object GetOptions(CancellationToken cancellationToken)
+        {
+            return _changeNamespaceService.GetChangeNamespaceOptions(
+                _moveToNamespaceAnalysisResult.Document,
+                _moveToNamespaceAnalysisResult.OriginalNamespace,
+                _moveToNamespaceAnalysisResult.Namespaces);
+        }
+
+        protected override async Task<IEnumerable<CodeActionOperation>> ComputeOperationsAsync(object options, CancellationToken cancellationToken)
+        {
+            IEnumerable<CodeActionOperation> operations = null;
+
+            if (options is MoveToNamespaceOptionsResult moveToNamespaceOptions && !moveToNamespaceOptions.IsCancelled)
+            {
+                var moveToNamespaceResult = await _changeNamespaceService.MoveToNamespaceAsync(
+                    _moveToNamespaceAnalysisResult,
+                    moveToNamespaceOptions.Namespace,
+                    cancellationToken).ConfigureAwait(false);
+
+                if (moveToNamespaceResult.Succeeded)
+                {
+                    operations = SpecializedCollections.SingletonEnumerable(new ApplyChangesOperation(moveToNamespaceResult.UpdatedSolution));
```

Comment:
Now that I think about this, I wonder if changing a type's namespace should trigger the VS refactor notify logic? Maybe then moving a WPF control in a .xaml.cs to a different namespace would magically fix up the XAML. Should probably try this.

Reply:
See: https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivsrefactornotify.onglobalsymbolrenamed?view=visualstudiosdk-2017

Reply:
Likely we'll need to trigger this for all code actions doing a namespace rename right? Let's file an issue outside this PR, there might be a more generic way to solve this.